docs/usage/operations.rst

```diff
@@ -723,3 +723,133 @@ Class Definitions
                    ),
               ),
           ]
+
+
+.. py:class:: SaferAddFieldOneToOne(model_name: str, name: str, field: models.OneToOneField)
+
+    Provides a safer way to add a one-to-one field to an existing model
+
+    :param model_name: Model name in lowercase without underscores.
+    :type model_name: str
+    :param name: The column name for the new one-to-one field.
+    :type name: str
+    :param field: The one-to-one field that is being added.
+    :type field: models.OneToOneField
+
+    **Why use this SaferAddFieldOneToOne operation?**
+    -------------------------------------------------
+
+    When using Django's default ``AddField`` operation, the SQL created has the
+    following form:
+
+    .. code-block:: sql
+
+      BEGIN;
+      --
+      -- Add field foo to bar
+      --
+      ALTER TABLE "myapp_bar"
+      ADD COLUMN "foo_id" bigint NULL
+      UNIQUE CONSTRAINT "auto_gen_constraint_name"
+      REFERENCES "myapp_foo"("id")
+      DEFERRABLE INITIALLY DEFERRED;
+
+      SET CONSTRAINTS "auto_gen_constraint_name" IMMEDIATE;
+      COMMIT;
+
+
+    The ``ALTER TABLE`` command takes an AccessExclusive lock, which is the
+    highest level of locking. It will block reads and writes on the table.
+    At the same time, Postgres will serially create the constraint while that
+    lock is held, which can potentially take a long time.
+
+    The below are the queries executed by this operation in order to avoid the
+    two problems above:
+
+    .. code-block:: sql
+
+      -- This operation takes an AccessExclusiveLock, but for a very short
+      -- duration. Adding a nullable field in Postgres doesn't require a full
+      -- table scan starting on version 11.
+      ALTER TABLE "myapp_bar" ADD COLUMN IF NOT EXISTS "foo_id" bigint NULL;
+
+      -- This operation takes an ShareUpdateExclusiveLock. It won't block
+      -- reads or writes on the table.
+      SET lock_timeout TO '0';
+      CREATE UNIQUE INDEX CONCURRENTLY IF NOT EXISTS "bar_foo_id_uniq" ON "myapp_bar" ("foo_id");
+      SET lock_timeout TO '10s';
+
+      -- This operation takes an AccessExclusiveLock, but for a very short
+      -- duration as it leverages the unique constraint index above to create
+      -- the constraint. By default this constraint will be IMMEDIATE, which
+      -- is what we want.
```

Comment:
❓  Why do we want the constraint to be IMMEDIATE?

Reply:
ooph that comment is not relevant at all. We can't alter the constraint unless
it's an fk constraint anyway, so it will always be immediate which makes the
comment irrelevant.

```
ALTER CONSTRAINT #

    This form alters the attributes of a constraint that was previously created. Currently only foreign key constraints may be altered.
```

74ab500da23a32d2d5a82b73d11328dd23cea7d0